### ISSUE TITLE
Constructor and prototype reform tracking

### ISSUE THREAD

**Issue by domenic** [jsdom/jsdom#2727]
Constructor and prototype reform has landed in https://github.com/jsdom/jsdom/commit/5e39a4c60377c95ac09ea938ecbf5f1f91cb0360, but not yet been released. This issue is meant as a meta tracking issue for remaining potential work items.

**Currently I do not think any of these should block the next release, and am considering doing the release some time in the next few days.** So please speak up if you disagree! I am worried a bit about how fixing these issues could be considered a breaking change, in which case we may want to bundle it with the existing constructor and prototype reform breaking change that has already landed.

### Classes from external packages are not yet migrated

Namely, `DOMException`, `URL`, `URLSearchParams`, and `XMLSerializer`.

@pmdartus is working on this, and it shouldn't be too hard, except he ran into the following issue while doing so...

### ECMAScript spec globals are handled inconsistently

Discussed in https://github.com/jsdom/domexception/pull/3#discussion_r349866219.

Currently, if `runScripts` is set to either `"outside-only"` or `"dangerously"`, the Node.js `vm` machinery will install all the ECMAScript globals onto the JSDOM `Window`. I.e., `dom.window.TypeError`, `dom.window.Map`, etc. will exist, and will be fresh copies: `dom.window.TypeError !== global.TypeError`. If `runScripts` is left as its default `undefined`, then these will not exist.

The exception is the typed arrays, which we [alias from the Node.js outer environment](https://github.com/jsdom/jsdom/blob/3e0655fc0651f233c90117eca108b66453d21247/lib/jsdom/browser/Window.js#L443-L452). We do this regardless of the `runScripts` value. Presumably this was to cause us to pass more tests in some way, but as noted in the comment, it's an inconsistent hack. I'd like to stop doing this, one way or another.

I see a few options here:

- Remove support for disabling scripting, i.e. make the new default equivalent to `"outside-only"`. And, remove the typed array special-casing. Now, fresh copies of all ECMAScript globals will be installed, just like fresh copies of web globals are now installed. This is nice for consistency, but may hurt performance in the simple case where people don't need to run scripts.

- Remove the aliasing of typed arrays entirely. Now, there will be no ECMAScript globals if scripting is disabled, which is at least somewhat more consistent. Presumably in this case any JSDOM code that wants to use the globals will use copies from the outer Node.js environment, at least when scripting is disabled. (E.g., `dom.window.DOMException` would inherit from `global.Error`, not from `dom.window.Error`, since the latter does not exist.)

- Alias all the ECMAScript globals when scripting is disabled (and remove the typed array aliasing for when scripting is enabled). Now at least there's a full set of globals at all times, although when scripting is disabled, it will be inconsistent where they come from between ES globals (aliases of the outer Node.js ones) and web ones (fresh).

Any of these options seem reasonable to me; thoughts appreciated.

### Errors and other globals from the outer Node.js environment are still used

For example, webidl2js-generated `TypeError`s, or manually-thrown ones inside impl classes, are thrown using `global.TypeError`, not `dom.window.TypeError`.

This also holds for other globals that we expose outward to the world. The notable ones I can think of are return values that are arrays or promises.

This is even more complicated because the exact realm in which these globals is created is not always the same. Specs need to explicitly state which of (at least) two choices they are using: the _relevant realm_, i.e. the global of `this`, or the _current realm_, i.e. the global of the currently running function. These would differ in cases like `dom1.window.SomeInterface.prototype.someOperation.call(new dom2.window.SomeInterface())`: there the relevant realm is that of `dom2` and the current realm is that of `dom1`.

Generally speaking, thrown errors always use the current realm, and promises created by the automatic Web IDL machinery will use the current realm, but most other cases will use the relevant realm.

One solution here is to update all of our code (including generated code) to use the relevant realm, e.g. changing `new TypeError()` to `new this._globalObject.TypeError()`. But since relevant is not current, that isn't really correct in cross-realm cases. The only way to get correct current-realm behavior, I believe, is discussed in the next section.

### Functions are not instanceof `dom.window.Function`

That is, `dom.window.Blob instanceof dom.window.Function` is false. Instead, `dom.window.Blob instanceof global.Function` is true. This causes at least one web platform test to fail.

Although you could imagine "reparenting" every function created with webidl2js via `Object.setPrototypeOf`, that seems pretty annoying.

An alternate route would be to consider mashing all of the generated wrapper files into a `vm.Script` and running them inside the VM, so that the classes in question are actually created inside the VM realm, instead of created outside and then attached to the VM instance. This gets pretty scary. For example: do we need to drag along the impl files to? Or maybe we consider smuggling them in through a different channel, e.g. global properties which are grabbed during startup and then deleted afterward? How does this interact with people who want to debug jsdom code---probably debugging a giant `vm.Script` is less good than debugging Node.js modules? Etc. This would also solve the previous problem though, as automatically all textual instances of `new TypeError()` would be interpreted as being done in the correct "current realm".

**Comment by Sebmaster** [comment]
> An alternate route would be to consider mashing all of the generated wrapper files into a vm.Script and running them inside the VM, so that the classes in question are actually created inside the VM realm, instead of created outside and then attached to the VM instance. This gets pretty scary. For example: do we need to drag along the impl files to?

I kinda like this a lot. I think we could just shove `require` into the vm which should allow regular breakpoints for impls then and use the regular require cache so we'd only have 1 instance of impls? We could also fix stack traces by setting the filename of the scripts properly, although that'd require separate script instances per interface.

However I do wonder what the performance of that may be. I think there's some kind of bytecode cache we may have to write/expose for writing?

**Comment by domenic** [comment]
Yeah, https://nodejs.org/api/vm.html#vm_script_createcacheddata may help.

**Comment by pmdartus** [comment]
> Classes from external packages are not yet migrated

I am almost done with the `DOMException` one. I don't think there will be any issues with the rest of the packages.

> Alias all the ECMAScript globals when scripting is disabled (and remove the typed array aliasing for when scripting is enabled). Now at least there's a full set of globals at all times, although when scripting is disabled, it will be inconsistent where they come from between ES globals (aliases of the outer Node.js ones) and web ones (fresh).

I would personally vote for this approach. Having a full set of globals at all times would greatly simplify jsdom internals. Even if the source of the globals is inconsistent depending if scripting is enabled or not, I don't think it will be visible from the consumer code perspective since no script would run.

**Comment by domenic** [comment]
> I would personally vote for this approach.

I'm OK going with this, at least for now. However:

> Even if the source of the globals is inconsistent depending if scripting is enabled or not, I don't think it will be visible from the consumer code perspective since no script would run.

It is visible from the outside, using code like the OP's. In this world `dom.window.TypeError === global.TypeError` and `dom.window.URL !== global.URL`.

**Comment by TimothyGu** [comment]
Just putting this here so that it doesn't get lost. It appears that https://github.com/jsdom/jsdom/blob/931d9d1f390c44d5eea66fca52ebb3f6dbbdb585/lib/jsdom/living/xmlhttprequest.js#L98-L99 should also be part of this issue.

**Comment by ExE-Boss** [comment]
@TimothyGu I’m&nbsp;doing&nbsp;that in&nbsp;<https://github.com/jsdom/jsdom/pull/2737>.

**Comment by ExE-Boss** [comment]
<https://github.com/jsdom/jsdom/pull/2884> converts `NodeFilter` to use **WebIDL**.

**Comment by domenic** [comment]
I believe this is also causing [navigator-pluginarray](https://github.com/web-platform-tests/wpt/blob/1277b8019d3085dd266266c792e87087a6376286/html/webappapis/system-state-and-capabilities/the-navigator-object/navigator-pluginarray.html) WPT to fail. We are installing the Node environment's `Array.prototype[Symbol.iterator]` which mismatches the VM environment's, so the assertion `assert_equals(navigator.plugins[Symbol.iterator], arrayIterator)` fails.

**Comment by ExE-Boss** [comment]
@domenic I’m fixing that in <https://github.com/jsdom/webidl2js/pull/234>.
